subcommand/get-consul-client-ca/command_test.go

```diff
@@ -0,0 +1,428 @@
+	require.NoError(t, err)
+	defer os.Remove(outputFile.Name())
+
+	caFile, certFile, keyFile, cleanup := generateServerCerts(t)
+	defer cleanup()
+
+	ui := cli.NewMockUi()
+	cmd := Command{
+		UI: ui,
+	}
+
+	// start the test server
+	a, err := testutil.NewTestServerConfigT(t, func(c *testutil.TestServerConfig) {
+		c.Connect = map[string]interface{}{
+			"enabled": true,
+		}
+		c.CAFile = caFile
+		c.CertFile = certFile
+		c.KeyFile = keyFile
+	})
+	require.NoError(t, err)
+	defer a.Stop()
+
+	// run the command
+	exitCode := cmd.Run([]string{
+		"-server-addr", strings.Split(a.HTTPSAddr, ":")[0],
+		"-server-port", strings.Split(a.HTTPSAddr, ":")[1],
+		"-ca-file", caFile,
+		"-output-file", outputFile.Name(),
+	})
+	require.Equal(t, 0, exitCode, ui.ErrorWriter.String())
+
+	client, err := api.NewClient(&api.Config{
+		Address: a.HTTPSAddr,
+		Scheme:  "https",
+		TLSConfig: api.TLSConfig{
+			CAFile: caFile,
+		},
+	})
+	require.NoError(t, err)
+
+	// get the actual root ca cert from consul so we can compare that
+	// with the command output
+	roots, _, err := client.Agent().ConnectCARoots(nil)
+	require.NoError(t, err)
+	require.NotNil(t, roots)
+	require.NotNil(t, roots.Roots)
+	require.Len(t, roots.Roots, 1)
+	require.True(t, roots.Roots[0].Active)
+	expectedCARoot := roots.Roots[0].RootCertPEM
+
+	// read the file contents
+	actualCARoot, err := ioutil.ReadFile(outputFile.Name())
+	require.NoError(t, err)
+	require.Equal(t, expectedCARoot, string(actualCARoot))
+}
+
+// Test that if the Consul server is not available at first,
+// we continue to poll it until it comes up.
+func TestRun_ConsulServerAvailableLater(t *testing.T) {
+	t.Parallel()
+	outputFile, err := ioutil.TempFile("", "ca")
+	require.NoError(t, err)
+	defer os.Remove(outputFile.Name())
+
+	caFile, certFile, keyFile, cleanup := generateServerCerts(t)
+	defer cleanup()
+
+	ui := cli.NewMockUi()
+	cmd := Command{
+		UI: ui,
+	}
+
+	randomPorts := freeport.MustTake(6)
+
+	// Start the consul agent asynchronously
+	var a *testutil.TestServer
+	go func() {
+		// start the test server after 100ms
+		time.Sleep(100 * time.Millisecond)
+		a, err = testutil.NewTestServerConfigT(t, func(c *testutil.TestServerConfig) {
+			c.Ports = &testutil.TestPortConfig{
+				DNS:     randomPorts[0],
+				HTTP:    randomPorts[1],
+				HTTPS:   randomPorts[2],
+				SerfLan: randomPorts[3],
+				SerfWan: randomPorts[4],
+				Server:  randomPorts[5],
+			}
+			c.Connect = map[string]interface{}{
+				"enabled": true,
+			}
+			c.CAFile = caFile
+			c.CertFile = certFile
+			c.KeyFile = keyFile
+		})
+		require.NoError(t, err)
+	}()
+
+	exitCode := cmd.Run([]string{
+		"-server-addr", "localhost",
+		"-server-port", fmt.Sprintf("%d", randomPorts[2]),
+		"-ca-file", caFile,
+		"-output-file", outputFile.Name(),
+	})
+	require.Equal(t, 0, exitCode)
+
+	// make sure a has been initialized by the time we call Stop()
+	retry.Run(t, func(r *retry.R) {
+		require.NotNil(r, a)
+	})
+	defer a.Stop()
```

Comment:
if `require.Equal(t, 0, exitCode)` fails then the server will never get stopped. Can you instead do
```
go func() { ... }
defer func() { if a != nil { a.Stop() }}
```

right after you spawn the goroutine?

Reply:
Good catch! ty!